dataset/dataset.py

```diff
@@ -56,26 +65,27 @@ def __del__(self):
 
 
 class IterableCorpusDataset(IterableDataset):
-  def __init__(self, file_path, batch_size, num_workers, start=0, transform=None):
-    self.file_path = file_path
+  def __init__(self, file_paths, lengths, batch_size, num_workers, K=2097152,
+               T=2, transform=None):
+    self.file_paths = file_paths
     self.batch_size = batch_size
     self.num_workers = num_workers
-    self.start = start
     self.transform = transform
+    self.lengths = lengths
+    lengths_sum = sum(lengths)
+    weights = [min(length, K) / lengths_sum for length in lengths]
+    weights = list(map(lambda x: math.pow(x, (1 / T)), weights))
+    tot_weights = sum(weights)
+    self.weights = list(map(lambda x: x / tot_weights, weights))
 
   def __iter__(self):
     worker_info = torch.utils.data.get_worker_info()
-    dataset_itr = gzip.open(self.file_path, "rb")
+    dataset_iterators = {}
+    for idx, path in enumerate(self.file_paths):
+        dataset_iterators[idx] = gzip.open(path, "rb")
     if worker_info is None:
-      dataset_itr = gzip.open(self.file_path, "rb")
-      for _ in range(self.start):
-        next(dataset_itr)
-      return TextSimpleIterator(dataset_itr, self.transform)
+      return TextSimpleIterator(dataset_iterators, self.lengths,
+                                self.weights, self.transform)
     else:
-      worker_id = worker_info.id
-      for _ in range(self.start):
-        next(dataset_itr)
-      for _ in range(self.batch_size * worker_id):
```

Comment:
These lines are needed. Same as my comment above, if multiple workers are used, you don't want them to overlap when reading files. So each worker start reading at a specific point in the file given its index.